src/renderer/base/renderer.cpp

```diff
@@ -64,44 +64,90 @@ Renderer::~Renderer()
 // - HRESULT S_OK, GDI error, Safe Math error, or state/argument errors.
 [[nodiscard]] HRESULT Renderer::PaintFrame()
 {
-    FOREACH_ENGINE(pEngine)
+    auto tries = maxRetriesForRenderEngine;
+    while (tries > 0)
     {
-        auto tries = maxRetriesForRenderEngine;
-        while (tries > 0)
+        if (_destructing)
         {
-            if (_destructing)
-            {
-                return S_FALSE;
-            }
+            return S_FALSE;
+        }
+
+        // BODGY: Optimally we would want to retry per engine, but that causes different
+        // problems (intermittent inconsistent states between text renderer and UIA output,
+        // not being able to lock the cursor location, etc.).
+        const auto hr = _PaintFrame();
+        if (SUCCEEDED(hr))
+        {
+            break;
+        }
+
+        LOG_HR_IF(hr, hr != E_PENDING);
 
-            const auto hr = _PaintFrameForEngine(pEngine);
-            if (SUCCEEDED(hr))
+        if (--tries == 0)
+        {
+            // Stop trying.
+            _pThread->DisablePainting();
+            if (_pfnRendererEnteredErrorState)
             {
-                break;
+                _pfnRendererEnteredErrorState();
             }
+            // If there's no callback, we still don't want to FAIL_FAST: the renderer going black
+            // isn't near as bad as the entire application aborting. We're a component. We shouldn't
+            // abort applications that host us.
+            return S_FALSE;
+        }
+
+        // Add a bit of backoff.
+        // Sleep 150ms, 300ms, 450ms before failing out and disabling the renderer.
+        Sleep(renderBackoffBaseTimeMilliseconds * (maxRetriesForRenderEngine - tries));
+    }
 
-            LOG_HR_IF(hr, hr != E_PENDING);
+    return S_OK;
+}
 
-            if (--tries == 0)
+[[nodiscard]] HRESULT Renderer::_PaintFrame() noexcept
+{
+    {
+        _pData->LockConsole();
+        auto unlock = wil::scope_exit([&]() {
+            _pData->UnlockConsole();
+        });
+
+        // Last chance check if anything scrolled without an explicit invalidate notification since the last frame.
+        _CheckViewportAndScroll();
+
+        if (_currentCursorOptions)

```

Comment:
okay so if I'm reading this right:
Every frame, once at the start (before any engines), 
* if the last frame had a cursor visible cursor
  * invalidate where that cursor was
* then get our new cursor state. (this may be nullopt, if the cursor isn't visible)

then paint each of the engines.